src/accounts.rs

```diff
@@ -136,10 +137,18 @@ impl AccountsDB {
         if tx.signatures.is_empty() && tx.fee != 0 {
             Err(BankError::MissingSignatureForFee)
         } else {
+            // Check for unique account keys
+            let mut unique_keys: Vec<Pubkey> = vec![];
```

Comment:
Nice catch! The other way to get in this situation is to use the indexes, which I guarded against here:

https://github.com/solana-labs/solana/blob/master/src/runtime.rs#L119

You can use that same function `has_duplicates` to avoid that allocating that `Vec` on every transaction. In fact, once we've made that assumption (that there's no duplicates), we can do the same "unsafe" trick that runtime.rs does to change `Self::load` to return a `&mut Account` instead of cloning the account. No need to do that in this PR - just pointing out the big optimization we can do as a consequence of this check!

Reply:
On second thought, we still need that clone in `Self::load` (or at least a `Cow`). If the program generates an error, we roll back by dropping that copy.